WordPress/Classes/System/WordPressAppDelegate+openURL.swift

```diff
@@ -0,0 +1,162 @@
+        let postId = params.number(forKey: "postId")
+
+        WPTabBarController.sharedInstance()?.showReaderTab(forPost: postId, onBlog: blogId)
+
+        return true
+    }
+
+    private func handleViewStats(url: URL) -> Bool {
+        guard let query = url.query,
+            let params = query.dictionaryFromQueryString(),
+            params.count > 0 else {
+                return false
+        }
+
+        let blogService = BlogService(managedObjectContext: ContextManager.sharedInstance().mainContext)
+
+        guard let siteId = params.number(forKey: "siteId"),
+            let blog = blogService.blog(byBlogId: siteId) else {
+            return false
+        }
+
+        let statsViewController = StatsViewController()
+        statsViewController.blog = blog
+        statsViewController.dismissBlock = {
+            WPTabBarController.sharedInstance()?.dismiss(animated: true, completion: nil)
+        }
+
+        let navController = UINavigationController(rootViewController: statsViewController)
+        navController.modalPresentationStyle = .currentContext
+        navController.navigationBar.isTranslucent = false
+
+        WPTabBarController.sharedInstance()?.present(navController, animated: true, completion: nil)
+
+        return true
+    }
+
+    @nonobjc private func handleDebugging(url: URL) {
+        guard let query = url.query,
+            let params = query.dictionaryFromQueryString(),
+            let debugType = params.string(forKey: "type"),
+            let debugKey = params.string(forKey: "key") else {
+                return
+        }
+
+        if debugKey == ApiCredentials.debuggingKey(), debugType == "crashlytics_crash" {
+            Crashlytics.sharedInstance().crash()
+        }
+    }
+
+    /// Handle a call of wordpress://newpost?…
+    ///
+    /// This is partially a return of the old functionality: https://github.com/wordpress-mobile/WordPress-iOS/blob/d89b7ec712be1f2e11fb1228089771a25f5587c5/WordPress/Classes/ViewRelated/System/WPTabBarController.m#L388
+    private func handleNewPost(url: URL) -> Bool {
+        guard let query = url.query,
+            let params = query.dictionaryFromQueryString() else {
+                return false
+        }
+
+        let title = params.string(forKey: NewPostKey.title)
+        let content = params.string(forKey: NewPostKey.content)
+        let tags = params.string(forKey: NewPostKey.tags)
+
+        // TODO: add ability to attach and image
+        //let image = params.string(forKey: NewPostKey.image)
+
+        let context = ContextManager.sharedInstance().mainContext
+        let blogService = BlogService(managedObjectContext: context)
+        guard let blog = blogService.lastUsedOrFirstBlog() else {
+            return false
+        }
+        let post = PostService(managedObjectContext: context).createDraftPost(for: blog)
+        post.postTitle = title
+        post.setPostFormatText(content ?? "no")
+        post.content = content
```

Comment:
Should we clean up the HTML and avoid malicious tags? What does Aztec render for a `<script>`? Or does it preserve attributes like `onclick` that could run malicious JS without the author noticing?

Reply:
@diegoreymendez didn't think it would be a problem for Aztec, but iirc like we're not completely sure how Gutenberg would deal with it.

I dunno if we should attempt to clean up the HTML, beyond what we might require for our editors. We could wreck HTML that the user desires as likely as we'd fix something questionable?

Reply:
I can't think of a scenario in which malicious html could cause a problem in Aztec itself, but my concern is about Aztec preserving malicious html/attributes that it won't render itself, but then would cause trouble when published.

What I'm thinking of is that WordPress generally trusts administrators and editors to publish [unfiltered_html](https://codex.wordpress.org/Roles_and_Capabilities#unfiltered_html), and otherwise use [wp_kses](https://codex.wordpress.org/Function_Reference/wp_kses) to filter out potentially malicious content.

What we need to ensure is that any content that comes from an external source is either marked unsafe somehow so KSES can do its thing on the backend (would require core and API changes), make that filtering happen on the app (big effort and hard to maintain), or at least make sure that any potentially unsafe content is apparent to the user, so they can inspect further and make a conscious decision about it.

I see for instance that Aztec will show `[SCRIPT]` if you just try to put a `<script>` tag into a post, which would be enough for the user to notice:

![img_f488d895f466-1](https://user-images.githubusercontent.com/8739/54023182-99094e80-4194-11e9-9c49-a729ab12027a.jpeg)

However, this is not the only way to inject JavaScript. You can add `onload="alert(1)"` to an `img` tag and Aztec will display the image without any hints to that attribute, and WordPress will not filter that out.

I don't know what the right solution is, but I can't think of a quick solution that doesn't create a vulnerability.

Reply:
Thanks for the explanation, @koke. It's a shame we didn't have tracks in place for the previous implementation so we could see how widely used it was.

A few 'low tech' suggestions:

* Don't allow HTML. Only allow text or Markdown, and we'll render the Markdown to HTML in the app? (we already do this in Simplenote) A bit limiting, but should still be usable from apps like Bear.
* Whitelist specific apps to be able to use the URL scheme, based on the `x-callback-url` `x-source` parameter. For example, apps like Bear or Ulysses aren't going to be sending anything anything to us automatically, and it should all be the user's content. This wouldn't be ideal, but still an improvement over having no URL scheme available.
* When we launch the editor from another app, we present an alert asking the user to confirm that they'd like to import content from X app. Another roadblock for the user to sanity check what they're importing.

Ultimately, I wonder how much of a concern this should be – it's a feature that I imagine will have relatively low use, and users still have to explicitly publish the post themselves. 🤔 

@matteorattotti Can I ask whether you have any thoughts on this? I know Bear has a URL scheme, although it doesn't publish that content anywhere.

Reply:
> Don't allow HTML. Only allow text or Markdown

Plain text is safe but quite limiting. Markdown would be safer than HTML, but we don't currently have any support for it. If we're considering doing the work of brining a markdown engine into the app we might as well do the work of accepting HTML and implement a KSES-like system to filter dangerous tags/attributes.

> Whitelist specific apps to be able to use the URL scheme

Unless you start adding cryptographic signatures (and you can't even protect the secrets for that on mobile), I think this would be trivial to bypass?

> We present an alert asking the user to confirm that they'd like to import content from X app

I'm not sure what this will add to the UX, app switching would already make that pretty obvious I think. Showing the raw html about to be imported could prevent surprises but it'd likely be a terrible UX in general.

Reply:
> You can add onload="alert(1)" to an img tag and Aztec will display the image without any hints to that attribute, and WordPress will not filter that out.

I spent some time today trying to make this happen, and I couldn't. It was filtered out both on `wpcom` and self-hosted sites. Is it possible to know if the page was filtered with `wp_kses`, or another method?

I added an example link to [the test page](
http://heagy.com/urltest) that you can try out with this branch.

Also, this appears to be the same way the current share extension works: the incoming text is used as the content for a page. If it happens to actually contain HTML, it renders as HTML perfectly fine. You can test this by sharing from the Notes app.

Please let me know if my testing is adequate, or if there is a reliable way to get suspicious code into a site on this branch (or via the share sheet 😬)

Reply:
> @matteorattotti Can I ask whether you have any thoughts on this? I know Bear has a URL scheme, although it doesn't publish that content anywhere.

We're seeing a lot of use for the URL scheme, mostly thanks for the Siri Shortcuts, I can imagine people creating posts by getting information from other apps using Siri  Shortcuts.

Using only markdown should be fine, but that's more an internal decision for you guys.

Reply:
> I spent some time today trying to make this happen, and I couldn't.

I tested your link (before the last commit that only accepts plain text) and It went through for me:

https://cloudup.com/cnwiThHEvW3

Reply:
I think we can revisit this in the future. For Gutenberg, this would still be problematic if you just create a new post and set the HTML, but the code that handles pasting on Gutenberg is already quite smart at filtering out bad things and converting things to blocks.

When we have support for a few more blocks and have Gutenberg as the default editor, I think we could start allowing HTML here, by creating an empty post and then "pasting" that HTML into it

Reply:
> It went through for me

@koke is that sandbox stand-alone self-hosted? I found that self-hosted sites without Jetpack will allow the javascript through. That prompted the change to escape HTML entities. Sorry I didn't update this thread.

> I think we can revisit this in the future. For Gutenberg, this would still be problematic if you just create a new post and set the HTML, but the code that handles pasting on Gutenberg is already quite smart

That sounds like a great solution to me 👍

Reply:
No, it's a self hosted with Jetpack: https://sandbox.koke.me